init: per-core init hook

ycsin opened the pull request:
Spinned-off from #65824, with an in-tree example of how it can be used.

npitre reviewed (commented):
I think calling `z_riscv_pmp_init())` deep from SOC code is odd and wrong.
That ought to remain out of the SOC code's concern.
Similarly for IRQs.

ycsin commented:
> I think calling `z_riscv_pmp_init())` deep from SOC code is odd and wrong. That ought to remain out of the SOC code's concern. Similarly for IRQs.

Reverted the implementation to align with #65824, is it better now?

nashif commented on arch/riscv/Kconfig:
```diff
@@ -112,6 +112,12 @@ config RISCV_SOC_HAS_CUSTOM_SYS_IO
 	  the RISC-V SoC needs to do something different and more than reading and
 	  writing the registers.
 
+config RISCV_SOC_HAS_CUSTOM_PER_CORE_INIT
```
this should not be riscv specific, the same pattern is use elsewhere, maybe such customization is used here first, but IMO this should be generalized to accomodate future usage by other arches, so I suggest doing this the same way we did it in `include/zephyr/platform/hooks.h` and use a similar pattern.

nashif commented on arch/riscv/core/smp.c:
```diff
@@ -28,6 +28,10 @@ extern void __start(void);
 #endif
 
+#ifdef CONFIG_RISCV_SOC_HAS_CUSTOM_PER_CORE_INIT
+void z_soc_per_core_init(void);
```
lets not add more `z_soc...` code into the tree, just drop the `z_ ` from the name, this can be confusing as  it describes something internal to zephyr that should not be used outside of zephyr, what you are doing here is defining a new "interface", so it is an API somehow.

ycsin committed:
init: support per-core init hook

Allow SoC to implement their custom per-core initialization function by
selecting `CONFIG_SOC_PER_CORE_INIT_HOOK` and implement
`soc_per_core_init_hook()`.

Signed-off-by: Maxim Adelman <imax@meta.com>
Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
soc: andestech: soc_per_core_init_hook() shouldn't return value

The `soc_per_core_init_hook()` function now has `void` type after
da118b9, so it should just return without any value.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
soc: andestech: run pma_init_per_core() with soc_per_core_init_hook()

The function `pma_init_per_core()`, as its name suggest, should be
run from every core, so call it from `soc_per_core_init_hook()`

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

evgeniy-paltsev reviewed (dismissed):
Looks reasonable. ACK for ARC part.

cfriedt commented:
What, no tests? 😁

jimmyzhe commented:
Thank you apply per core init hook on ae350 PMA feature , but I got some error when I tested it on adp_xc7k/ae350 with `CONFIG_SOC_ANDES_V5_PMA=y`.

```
/home/jimmyzhe/zephyrproject-upstream/zephyr/arch/riscv/include/kernel_arch_func.h: In function 'arch_kernel_init':
/home/jimmyzhe/zephyrproject-upstream/zephyr/arch/riscv/include/kernel_arch_func.h:57:9: error: implicit declaration of function 'soc_per_core_init_hook' [-Werror=implicit-function-declaration]
   57 |         soc_per_core_init_hook();
      |         ^~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

I think all file in 053112fe9b3ffbc426898f3d8f50b7bed58eab33 need to include zephyr/platform/hooks.h.

ycsin committed:
soc: andestech: refactor out soc_early_init_hook() from pma.c

Refactor out the `soc_early_init_hook()` function from `pma.c` to
`soc.c` which is always compiled so that it can be extended to run
other init functions easily in the future. Then, restore the function
in `pma.c` to `pma_init()`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
> Thank you for applying per core init hook on ae350 PMA feature , but I got error when I tested it on adp_xc7k/ae350 with `CONFIG_SOC_ANDES_V5_PMA=y`.
> 
> ```
> /home/jimmyzhe/zephyrproject-upstream/zephyr/arch/riscv/include/kernel_arch_func.h: In function 'arch_kernel_init':
> /home/jimmyzhe/zephyrproject-upstream/zephyr/arch/riscv/include/kernel_arch_func.h:57:9: error: implicit declaration of function 'soc_per_core_init_hook' [-Werror=implicit-function-declaration]
>    57 |         soc_per_core_init_hook();
>       |         ^~~~~~~~~~~~~~~~~~~~~~
> cc1: all warnings being treated as errors
> ```
> 
> I think all file in [053112f](https://github.com/zephyrproject-rtos/zephyr/commit/053112fe9b3ffbc426898f3d8f50b7bed58eab33) need to include zephyr/platform/hooks.h.

Oops, good catch. I also refactored the soc_*_init_hook()` functions from `pma.c` to `soc.c`
